packages/auto-id/src/utils.ts

```diff
@@ -26,8 +39,278 @@ export function derEncodeSignatureAlgorithmOID(
+  let identifier: string | null = null
+  const receipt: SubmittableResult = await new Promise((resolve, reject) => {
+    tx.signAndSend(signer, async (result: SubmittableResult) => {
+      const { events = [], status, dispatchError } = result
+
+      if (status.isInBlock || status.isFinalized) {
+        const blockHash = status.isInBlock
+          ? status.asInBlock.toString()
+          : status.asFinalized.toString()
+
+        try {
+          // Retrieve the block using the hash to get the block number
+          const signedBlock = await api.rpc.chain.getBlock(blockHash)
+          let identifier: string | null = null
+          events.forEach(({ event: { section, method, data } }) => {
+            if (section === 'system' && method === 'ExtrinsicFailed') {
+              const dispatchErrorJson = JSON.parse(dispatchError!.toString())
+
+              reject(
+                new Error(
+                  `Extrinsic failed: ${mapErrorCodeToEnum(dispatchErrorJson.module.error)} in block #${signedBlock.block.header.number.toString()}`,
+                ),
+              )
+            }
+            if (section === 'autoId' && method === 'NewAutoIdRegistered') {
+              identifier = data[0].toString()
+            }
+          })
+          resolve(result)
+        } catch (err: any) {
+          reject(new Error(`Failed to retrieve block information: ${err.message}`))
+        }
+      } else if (status.isDropped || status.isInvalid) {
+        reject(new Error('Transaction dropped or invalid'))
+      }
+    })
+  })
+
+  return { receipt, identifier }
+}
+
+export const identifierFromX509Cert = (
+  issuerId: string | null | undefined,
+  certificate: X509Certificate,
+): string => {
+  const subjectCommonName = CertificateManager.getSubjectCommonName(certificate.subjectName)
+  if (!subjectCommonName) {
+    throw new Error('Subject Common Name not found')
+  }
+
+  // convert from string to bytes.
+  const subjectCommonNameBytes = new TextEncoder().encode(subjectCommonName)
+
+  if (issuerId) {
+    const issuerIdBytes = new TextEncoder().encode(issuerId)
+    const data = new Uint8Array(issuerIdBytes.length + subjectCommonNameBytes.length)
+    data.set(issuerIdBytes)
+    data.set(subjectCommonNameBytes, issuerIdBytes.length)
+
+    return blake2b_256(data)
+  } else {
+    return blake2b_256(subjectCommonNameBytes)
+  }
+}
+
+// Utility function to prepare signing data
+export const prepareSigningData = async (
+  api: ApiPromise,
+  autoIdIdentifier: string,
+  getCertificate: (id: string) => Promise<AutoIdX509Certificate | null>,
+  actionType: CertificateActionType,
+): Promise<{
+  serializedData: Uint8Array
+  algorithmOid: AsnAlgorithmIdentifier
+}> => {
+  const certificate = await checkCertificateAndRevocationList(api, autoIdIdentifier, getCertificate)
+  const subjectPublicKeyInfo = certificate.subjectPublicKeyInfo
+  const algorithmOid = extractSignatureAlgorithmOID(subjectPublicKeyInfo)
```

Comment:
I thought we decided to not extract the signing key from the certificate since this may not work because `subjetPublicKeyInfo` contains the key generating algorithm, not necessarily the signing algorithm used for the auto-id functions. This likely why you are having problems with [RSA signing](https://github.com/subspace/auto-sdk/pull/75#discussion_r1684172259).

Reply:
Hmm...Understood!

Now it all makes sense to me.
That's why in RSA, I saw 2 OIDs: one for key generating & other for signing algorithms. And that's why the error. But, that's not the case with Ed25519.

Thanks!